services/src/main/java/io/grpc/services/HealthServiceImpl.java

```diff
@@ -16,30 +16,40 @@
 
 package io.grpc.services;
 
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.util.concurrent.MoreExecutors;
+import io.grpc.Context.CancellationListener;
+import io.grpc.Context;
 import io.grpc.Status;
 import io.grpc.StatusException;
 import io.grpc.health.v1.HealthCheckRequest;
 import io.grpc.health.v1.HealthCheckResponse;
 import io.grpc.health.v1.HealthCheckResponse.ServingStatus;
 import io.grpc.health.v1.HealthGrpc;
 import io.grpc.stub.StreamObserver;
-import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import javax.annotation.Nullable;
-
+import javax.annotation.concurrent.GuardedBy;
 
 final class HealthServiceImpl extends HealthGrpc.HealthImplBase {
 
   /* Due to the latency of rpc calls, synchronization of the map does not help with consistency.
-   * However, need use ConcurrentHashMap to prevent the possible race condition of concurrently
-   * putting two keys with a colliding hashCode into the map.*/
-  private final Map<String, ServingStatus> statusMap
-      = new ConcurrentHashMap<String, ServingStatus>();
+   * However, need use ConcurrentHashMap to allow concurrent reading by check() */
+  private final ConcurrentHashMap<String, ServingStatus> statusMap =
+      new ConcurrentHashMap<String, ServingStatus>();
+
+  private final Object watchLock = new Object();
+
+  @GuardedBy("watchLock")
+  private final Multimap<String, StreamObserver<HealthCheckResponse>> watchers =
```

Comment:
If possible I think it would be good if this was a Map<String, List<StreamObserver>>, to avoid guava deps

Reply:
Although Guava collections dependency is permissible for server-side code, making Multimap use identity hash set would require extra code and lose the benefits of being shorter. I changed to the plain Java collections.